discovery-provider/src/queries/get_tips.py

```diff
@@ -0,0 +1,286 @@
+    offset: int
+    user_id: Optional[int]
+    receiver_min_followers: Optional[int]
+    receiver_is_verified: Optional[bool]
+    current_user_follows: Optional[str]
+    unique_by: Optional[str]
+    min_slot: int
+    transaction_signatures: List[int]
+
+
+class TipResult(TypedDict):
+    amount: int
+    sender: str
+    receiver: str
+    slot: int
+    created_at: datetime
+    followee_supporters: List[str]
+    transaction_signature: str
+
+
+# Example of query with inputs:
+# limit=100, offset=0, user_id=1, current_user_follows=sender_or_receiver
+# ---------------------------------------------------------
+# WITH followees AS
+# (
+#     SELECT
+#         follows.followee_user_id AS followee_user_id
+#     FROM
+#         follows
+#     WHERE
+#         follows.is_current = true
+#         AND follows.is_delete = false
+#         AND follows.follower_user_id = 1
+# )
+# ,
+# tips AS
+# (
+#     SELECT
+#         user_tips.signature AS signature,
+#         user_tips.slot AS slot,
+#         user_tips.sender_user_id AS sender_user_id,
+#         user_tips.receiver_user_id AS receiver_user_id,
+#         user_tips.amount AS amount,
+#         user_tips.created_at AS created_at
+#     FROM
+#         user_tips
+#         LEFT OUTER JOIN
+#             followees AS followees_for_receiver
+#             ON user_tips.receiver_user_id = followees_for_receiver.followee_user_id
+#         LEFT OUTER JOIN
+#             followees AS followees_for_sender
+#             ON user_tips.sender_user_id = followees_for_sender.followee_user_id
+#     WHERE
+#         followees_for_receiver.followee_user_id IS NOT NULL
+#         OR followees_for_sender.followee_user_id IS NOT NULL
+#     ORDER BY
+#         user_tips.slot DESC LIMIT 100 OFFSET 0
+# )
+# ,
+# followee_tippers AS
+# (
+#     SELECT
+#         aggregate_user_tips.sender_user_id AS sender_user_id,
+#         aggregate_user_tips.receiver_user_id AS receiver_user_id,
+#         followees_for_aggregate.followee_user_id AS followee_user_id
+#     FROM
+#         followees AS followees_for_aggregate
+#         LEFT OUTER JOIN
+#             aggregate_user_tips
+#             ON aggregate_user_tips.sender_user_id = followees_for_aggregate.followee_user_id
+# )
+# SELECT
+#     tips.signature AS tips_signature,
+#     tips.slot AS tips_slot,
+#     tips.sender_user_id AS tips_sender_user_id,
+#     tips.receiver_user_id AS tips_receiver_user_id,
+#     tips.amount AS tips_amount,
+#     tips.created_at AS tips_created_at,
+#     array_agg(followee_tippers.sender_user_id) AS array_agg_1
+# FROM
+#     tips
+#     LEFT OUTER JOIN
+#         followee_tippers
+#         ON followee_tippers.receiver_user_id = tips.receiver_user_id
+# GROUP BY
+#     tips.signature,
+#     tips.slot,
+#     tips.sender_user_id,
+#     tips.receiver_user_id,
+#     tips.amount,
+#     tips.created_at
+# ORDER BY
+#     tips.slot DESC LIMIT 100 OFFSET 0;
+#
+
+
+def _get_tips(session: Session, args: GetTipsArgs):
```

Comment:
@isaacsolo would you lend your SQL expertise to this? Trying to make this efficient as possible.

The big timesinks seemed to be getting unique senders or receivers, filtering to users you follow, and including followers of yours who also have tipped the receiver. I added a bunch of CTE logic etc to try to speed this up - curious on your thoughts.

I have an example of what the SQL output looks like for some subset of the args above

Reply:
I know I merged this in, but any ideas you might have would still be valuable - can do a quick fast follow!

Reply:
@rickyrombo @piazzatron sorry for the late review! thanks for the sql above. btw i know there was some back/forth on ORM vs SQL. i think doing this with ORM and also having to include SQL is a good example of why reading SQL tends to be easier.

Reply:
The issue with just using SQL here is how conditional the query is - using ORM to build it makes sense I think in this case, but yeah I do like the readability of SQL in comparison especially when reviewing perf. Hopefully this feels like a best of both worlds!